File: tools/capture_scan_point_reference.m
function capture_scan_point_reference(out_json, list_fn)
% capture_scan_point_reference  Engine-free per-point byte capture for the W6 oracle.
%
% For each (scan, sequence) pair from scan_point_list, expands the ScanGroup point by
% point and serializes the sequence built for each point -- exactly the RunScans inner
% loop -- writing the per-point bytes (hex) to <...>/scan_point_reference.json:
%   { "<name>": { "seq": "<SeqFn>", "nseq": N, "points": ["<hex>", ...] }, ... }
%
% SAFETY (same as capture_ybseqs_reference): serialize() never runs the deferred
% regBeforeStart/regAfterEnd callbacks (camera/AWG/server/MemoryMap), and
% override_tick_per_sec keeps the libnacs engine unloaded. Real sequence code, so run it
% in a fresh headless `matlab -batch` and preferably a maintenance window:
%   matlab -batch "cd <pyctrl/tools>; capture_scan_point_reference"

    here = fileparts(mfilename('fullpath'));
    repo = fullfile(here, '..', '..');
    mn = fullfile(repo, 'matlab_new');
    if nargin < 1 || isempty(out_json)
        out_json = fullfile(here, '..', 'tests', 'reference_scan_point', ...
                            'scan_point_reference.json');
    end
    if nargin < 2 || isempty(list_fn)
        list_fn = @scan_point_list;
    end
    odir = fileparts(out_json);
    if ~exist(odir, 'dir'), mkdir(odir); end

    for d = {mn, fullfile(mn,'lib'), fullfile(mn,'YbSeqs'), fullfile(mn,'YbSteps'), ...
             fullfile(mn,'YbScans'), fullfile(mn,'YbScans','scanConfig'), ...
             fullfile(mn,'YbRearrangement'), fullfile(mn,'YbFPGAWaveforms'), ...
             fullfile(mn,'YbExptCtrl')}
        if exist(d{1}, 'dir'), addpath(d{1}); end
    end

    % Real production tick rate (1e12 == 1 ps): us-scale steps round to 0 ticks at the
    % default 1000/s and are rejected. Engine stays unloaded.
    SeqManager.override_tick_per_sec(1e12);
    cleanup = onCleanup(@() SeqManager.override_tick_per_sec(0)); %#ok<NASGU>

    specs = list_fn();
    out = struct();
    for i = 1:numel(specs)
        name = specs(i).name;
        g = specs(i).build();
        seqfn = str2func(specs(i).seq);
        ns = nseq(g);
        pts = cell(1, ns);
        for n = 1:ns
            params = getseq(g, n);              % resolved scalar params for point n
            s = ExpSeq(params);                 % merge params onto consts
            s = seqfn(s);                       % build the step tree (no callbacks run)
            b = serialize(s);                   % NO generate()/run()
            pts{n} = lower(sprintf('%02x', typecast(int8(b), 'uint8')));
        end
        entry = struct();
        entry.seq = specs(i).seq;
        entry.nseq = ns;
        entry.points = pts;                     % cell of hex -> JSON array of strings
        out.(name) = entry;
        fprintf('captured %s (%s): %d points\n', name, specs(i).seq, ns);
    end

    fid = fopen(out_json, 'w');
    fwrite(fid, jsonencode(out), 'char');
    fclose(fid);
    fprintf('done: %d (scan, seq) pairs -> %s\n', numel(specs), out_json);
end

File: tools/capture_scan_point_reference.py
"""Re-capture ``tests/reference_scan_point/scan_point_reference.json`` FROM pyctrl.

Pyctrl-side twin of the MATLAB ``capture_scan_point_reference.m``. Same rationale as
``capture_ybseqs_reference.py``: pyctrl is the authoritative runtime, so the strict
pyctrl<->MATLAB per-point byte-equality is RELEASED and this captures pyctrl's OWN per-point
``serialize()`` output as the regression reference. ``test_scan_point_oracle.py`` and
``test_dispatch_descriptor.py`` (which share this file) then become pyctrl self-consistency
guards. Engine-free (override_tick_per_sec), exactly as the tests build.

The scan builders MUST mirror ``tests/test_scan_point_oracle.py`` (same swept values) so the
captured bytes equal what the tests build.

    python pyctrl/tools/capture_scan_point_reference.py
"""

import json
import os
import sys

_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))   # .../pyctrl
for _p in (os.path.join(_ROOT, "lib"), os.path.join(_ROOT, "YbSteps"),
           os.path.join(_ROOT, "YbSeqs"), os.path.join(_ROOT, "YbExptCtrl"),
           os.path.join(_ROOT, "tools"), _ROOT):   # mirror pyproject.toml pythonpath
    if _p not in sys.path:
        sys.path.insert(0, _p)

import seq_manager                       # noqa: E402
from exp_seq import ExpSeq               # noqa: E402
from scan_group import ScanGroup         # noqa: E402
from seq_config import SeqConfig         # noqa: E402

_REF = os.path.join(_ROOT, "tests", "reference_scan_point", "scan_point_reference.json")


# --- twin builders of tests/test_scan_point_oracle.py (same scanned values) --------------- #
def build_spectrum399():
    g = ScanGroup()
    g().Pushout.Blue.Amp1 = 0.25
    g().Pushout.Blue.Freq.scan(1, [v * 1e6 for v in range(220, 361, 35)])  # 5 points
    g().Pushout.Time = 10e-3
    g.runp().NumPerGroup = 10000
    g.runp().NumImages = 2
    g.runp().Scramble = 1
    return g


def build_imaging_hist():
    g = ScanGroup()
    g().Imag399.ExposureTime = 100e-3
    g().SLM.VServo = 1
    g().Imag399.FreqDetuning.scan(1, [-5 * 1e6, 0 * 1e6])
    g().Imag399.Amp.scan(2, [0.2, 0.3])
    g().Pushout.Green.Amp = 0
    g().Pushout.Blue.Amp1 = 0
    g().Pushout.Time = 10e-3
    g.runp().NumImages = 2
    g.runp().Scramble = 1
    return g


_PAIRS = {
    "spectrum399": (build_spectrum399, "PushoutSurvival399Seq"),
    "imaging_hist": (build_imaging_hist, "PushoutSurvivalSeq"),
}


def main():
    SeqConfig.reset()
    SeqConfig.load_real()
    seq_manager.override_tick_per_sec(1e12)
    out = {}
    try:
        for name, (build, seqname) in _PAIRS.items():
            g = build()
            mod = __import__(seqname)
            seqfn = getattr(mod, seqname)
            points = [seqfn(ExpSeq(g.getseq(n))).serialize().hex()
                      for n in range(1, g.nseq() + 1)]
            out[name] = {"seq": seqname, "nseq": g.nseq(), "points": points}
    finally:
        seq_manager.override_tick_per_sec(0)
        SeqConfig.reset()

    with open(_REF, "w") as f:
        json.dump(out, f)
    print("re-captured %d scans (%s) from pyctrl -> %s"
          % (len(out), ", ".join("%s:%d" % (k, v["nseq"]) for k, v in out.items()), _REF))


if __name__ == "__main__":
    main()
